inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,368 @@
+        # Arrange
+        parameters = ApplicationAddSourceParameters(gpg_key_name=gpg_key_name, gpg_key_uri=gpg_key_uri, source_list_file_name=source_list_file_name, sources=sources)
+
+        # Act
+        mock_application_source_manager.add(parameters)
+
+        # Assert
+        if gpg_key_name and gpg_key_uri:
+            mock_verify_source.assert_called_once_with(source=gpg_key_uri.rsplit('/', 1)[0], dispatcher_broker=mock_application_source_manager._dispatcher_broker)
+            mock_add_gpg_key.assert_called_once_with(gpg_key_uri, gpg_key_name)
+        mock_create_file_with_contents.assert_called_once_with(f'/etc/apt/sources.list.d/{source_list_file_name}', sources)
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerList:
+    @pytest.mark.parametrize('file_content, expected_output', [
+        # Test with one source
+        (["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted"], 
+         [ApplicationSourceList(name="sources.list", sources=["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted"])]),
+        # Test with multiple sources
+        (["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted", "deb-src http://archive.ubuntu.com/ubuntu/ bionic main restricted"], 
+         [ApplicationSourceList(name="sources.list", sources=["deb http://archive.ubuntu.com/ubuntu/ bionic main restricted", "deb-src http://archive.ubuntu.com/ubuntu/ bionic main restricted"])]),
+        # Test with no sources
+        ([""], [ApplicationSourceList(name="sources.list", sources=[])]),
+        # Test with commented lines
+        (["# deb http://archive.ubuntu.com/ubuntu/ bionic main restricted"], [ApplicationSourceList(name="sources.list", sources=[])]),
+        # Test with blank lines
+        ([""], [ApplicationSourceList(name="sources.list", sources=[])]),
+    ])
+    @patch("glob.glob")
+    @patch("builtins.open", new_callable=mock_open)
+    def test_list(self, mock_open, mock_glob, file_content, expected_output):
+        # Mock the glob.glob call to return a list with one file
+        mock_glob.return_value = ["sources.list"]
+        # Mock the open call to return an iterator over the file_content list
+        handle = mock_open.return_value.__enter__.return_value
+        handle.__iter__.return_value = iter(file_content)
+        handle.readlines.return_value = file_content
+        # Create an instance of UbuntuApplicationSourceManager
+        manager = UbuntuApplicationSourceManager(None)
+        # Call the list method and assert that the output matches the expected output
+        assert manager.list() == expected_output
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerRemove
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerRemove:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)

<h3>Class: UbuntuApplicationSourceManager |Function: remove</h3>

[Source Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py?plain=1#L149)

[Unit Test Case Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py?plain=1#L89)

<h4>Test Case Scenarios:</h4>
The unit test `test_remove` in the class `TestUbuntuApplicationSourceManagerRemove` generates four test cases based on the parameters provided to the `@pytest.mark.parametrize` decorator. Here are the test cases along with their explanations:

1. Test Case: `test_remove[True-True-None]`
   - Explanation: This test case checks the scenario where both the source file and the GPG key exist, and no exception is expected. The `remove` method should successfully remove the source file and the GPG key.

2. Test Case: `test_remove[True-False-None]`
   - Explanation: This test case checks the scenario where the source file exists but the GPG key does not, and no exception is expected. The `remove` method should successfully remove the source file and log a warning about the missing GPG key.

3. Test Case: `test_remove[False-True-SourceError]`
   - Explanation: This test case checks the scenario where the source file does not exist but the GPG key does, and a `SourceError` exception is expected. The `remove` method should raise a `SourceError` when it tries to remove the non-existent source file.

4. Test Case: `test_remove[False-False-SourceError]`
   - Explanation: This test case checks the scenario where neither the source file nor the GPG key exist, and a `SourceError` exception is expected. The `remove` method should raise a `SourceError` when it tries to remove the non-existent source file, and log a warning about the missing GPG key.

In each test case, the `remove` method is called with a `parameters` fixture, which is an instance of `ApplicationRemoveSourceParameters` with `source_list_file_name='test.list'` and `gpg_key_name='test.key'`. The `remove_file`, `get_canonical_representation_of_path`, and `os.path.join` functions are mocked to control their behavior and isolate the `remove` method for testing.


<h4>Test Execution Results:</h4>

| Total Cases    | 4  |
| -------------- | -- |
| Passed Cases   | 4  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)